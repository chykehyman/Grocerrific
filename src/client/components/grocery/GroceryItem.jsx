import React from 'react';
import { shape, string, bool } from 'prop-types';

const propTypes = {
  item: shape({
    name: string.isRequired,
    purchaseStatus: bool.isRequired
  }).isRequired
};

const GroceryItem = ({ item: { name, purchaseStatus } }) => {
  let listWrapper = 'list-unstyled';
  if (purchaseStatus) {
    listWrapper += ' strike-through';
  }
  return (
    <tr>
      <td>&nbsp;</td>
      <td className={listWrapper}>{name}</td>
      <td>
        <button
          type="button"
          className="btn btn-primary"
        >
          Buy
        </button>
      </td>
      <td>
        <button
          type="button"
          className="btn btn-outline-danger btn-sm">
          Delete
        </button>
      </td>
    </tr>
  );
};

GroceryItem.propTypes = propTypes;

export default GroceryItem;
