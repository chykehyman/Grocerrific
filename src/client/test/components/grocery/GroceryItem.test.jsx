import GroceryItem from '../../../components/grocery/GroceryItem';

describe('<GroceryItem /> component', () => {
  const setup = () => {
    const props = {
      item: {
        _id: 'myID123',
        name: 'tomato',
        purchaseStatus: false
      },
      isProcessing: false,
      selectedId: 'myID123',
      action: '',
      deleteItem: jest.fn(),
      purchaseOrDropItem: jest.fn()
    };

    return {
      shallowWrapper: shallow(<GroceryItem {...props} />),
      props
    };
  };


  it('should render without crashing', () => {
    const { shallowWrapper } = setup();

    expect(toJson(shallowWrapper)).toMatchSnapshot();
  });

  it('should show a strike-through class when purchaseStatus is set to true', () => {
    const { shallowWrapper, props } = setup();

    shallowWrapper.setProps({
      ...props,
      item: {
        ...props.item, purchaseStatus: true
      }
    });

    expect(shallowWrapper.find('.strike-through')).toHaveLength(1);
    expect(shallowWrapper.find('button').first().text()).toEqual('Drop');
  });

  it('should invoke the purchaseOrDropItem method while showing a <Loader /> component', () => {
    const { shallowWrapper, props } = setup();

    shallowWrapper.find('button').first().simulate('click', {});

    shallowWrapper.setProps({
      ...props,
      isProcessing: true,
      action: 'purchase or drop'
    });

    expect(props.purchaseOrDropItem).toHaveBeenCalledWith(props.item._id);
    expect(shallowWrapper.find('MDSpinner')).toHaveLength(1);
    expect(shallowWrapper.find('button').first().text()).toEqual('Buy');
  });

  it('should invoke the deleteItem method while showing a <Loader /> component', () => {
    const { shallowWrapper, props } = setup();

    shallowWrapper.find('button').last().simulate('click', {});

    shallowWrapper.setProps({
      ...props,
      isProcessing: true,
      action: 'delete'
    });

    expect(props.deleteItem).toHaveBeenCalledWith(props.item._id);
    expect(shallowWrapper.find('MDSpinner')).toHaveLength(1);
    expect(shallowWrapper.find('button').first().text()).toEqual('Buy');
  });
});
