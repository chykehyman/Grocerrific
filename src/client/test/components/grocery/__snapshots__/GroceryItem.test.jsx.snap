// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<GroceryItem /> component should render without crashing 1`] = `
<tr>
  <td>
    Â 
  </td>
  <td
    className="list-unstyled"
  >
    tomato
  </td>
  <td>
    <button
      className="btn btn-sm btn-primary"
      disabled={false}
      onClick={[Function]}
      type="button"
    >
      Buy
    </button>
    <span
      className="pl-2"
    />
  </td>
  <td>
    <button
      className="btn btn-outline-danger btn-sm"
      disabled={false}
      onClick={[Function]}
      type="button"
    >
      Delete
    </button>
    <span
      className="pl-2"
    />
  </td>
</tr>
`;
